sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +135,189 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
 	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
-	unsigned int val_l, val_r, gain_l_val, gain_r_val, loop_cnt, i, reg;
+	unsigned int gain_val, i;
+	const unsigned int loop_cnt = 2;
 	int ret;
 
-	if (strstr(ucontrol->id.name, "FU02 Capture Volume") ||
-		strstr(ucontrol->id.name, "FU06 Capture Volume"))
-		loop_cnt = 2;
-	else if (strstr(ucontrol->id.name, "FU0E Boost") ||
-		strstr(ucontrol->id.name, "FU0C Boost"))
-		loop_cnt = 4;
-	else
-		loop_cnt = 1;
-
-	/* control value to 2s complement */
-	/* L channel */
-	gain_l_val = ucontrol->value.integer.value[0];
-	if (gain_l_val > mc->max)
-		gain_l_val = mc->max;
-	val_l = gain_l_val;
-
-	if (mc->shift == 8) {
-		gain_l_val = (gain_l_val * 10) << mc->shift;
-	} else {
-		gain_l_val =
-			((abs(gain_l_val - mc->shift) * RT715_SDCA_DB_STEP) << 8) / 1000;
-		if (val_l <= mc->shift) {
-			gain_l_val = ~gain_l_val;
-			gain_l_val += 1;
-		}
-		gain_l_val &= 0xffff;
-	}
-
-	/* R channel */
-	gain_r_val = ucontrol->value.integer.value[1];
-	if (gain_r_val > mc->max)
-		gain_r_val = mc->max;
-	val_r = gain_r_val;
-
-	if (mc->shift == 8) {
-		gain_r_val = (gain_r_val * 10) << mc->shift;
-	} else {
-		gain_r_val =
-			((abs(gain_r_val - mc->shift) * RT715_SDCA_DB_STEP) << 8) / 1000;
-		if (val_r <= mc->shift) {
-			gain_r_val = ~gain_r_val;
-			gain_r_val += 1;
-		}
-		gain_r_val &= 0xffff;
-	}
-
-	/* Lch*/
 	for (i = 0; i < loop_cnt; i++) {
-		ret = regmap_write(rt715->mbq_regmap, mc->reg + i * 2, gain_l_val);
+		gain_val =
+			rt715_sdca_vol_gain(ucontrol->value.integer.value[i], mc->max,
+				mc->shift);
+		ret = regmap_write(rt715->mbq_regmap, mc->reg + i, gain_val);
 		if (ret != 0) {
 			dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
-				mc->reg + i * 2, gain_l_val);
+				mc->reg + i, gain_val);
 			return ret;
 		}
 	}
 
-	/* Rch */
-	for (i = 0; i < loop_cnt; i++) {
-		reg = (i == 3) ? (mc->rreg - 2) | BIT(15) : mc->rreg + i * 2;
-		ret = regmap_write(rt715->mbq_regmap, reg, gain_r_val);
+	return 0;
+}
+
+static int rt715_sdca_set_amp_gain_4ch_put(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	const unsigned int gain_sft = 0x2f, vol_max = 0x7f;
+	unsigned int reg_base = kcontrol->private_value;
+	unsigned int gain_val, i;
+	int ret;
+
+	for (i = 0; i < 4; i++) {
```

Comment:
yes, hard-coding 4 in `i < 4` looks perfectly valid and easy to understand to me, so maybe we could remove `loop_cnt` above too?